File: clientPython/packet.py
import base64
import binascii
import hashlib
import hmac
import ipaddress
import struct
import time


def _ip_to_128bit(ip_str: str) -> int:
    ip_obj = ipaddress.ip_address(ip_str)
    if isinstance(ip_obj, ipaddress.IPv6Address):
        return int(ip_obj)
    return int(ip_obj) + 0xffff00000000  # IPv4-mapped: ::ffff:a.b.c.d


def build_seed_packet(sequence: int, seed: str, source_ip: str) -> bytes:
    current_time = int(time.time()) & 0xFFFFFFFF
    time_bytes = struct.pack(">I", current_time)

    mac = hmac.new(seed.encode(), time_bytes, hashlib.sha256)
    time_hash = mac.digest()[:8]

    ip_u128 = _ip_to_128bit(source_ip)
    mac_ip = hmac.new(seed.encode(), digestmod=hashlib.sha256)
    mac_ip.update(time_bytes)
    mac_ip.update(ip_u128.to_bytes(16, byteorder='big'))
    ip_hash = mac_ip.digest()[:4]

    return struct.pack('!4sI4s8s4s', b'RP01', sequence, time_bytes, time_hash, ip_hash)


def build_token_packet(sequence: int, token: bytes) -> bytes:
    return struct.pack('!4sI16s', b'RP01', sequence, token)


def decode_token(token_b64: str) -> bytes:
    try:
        token = base64.b64decode(token_b64)
    except (ValueError, binascii.Error) as e:
        raise ValueError(f"invalid base64 token: {e}") from e
    if len(token) != 16:
        raise ValueError("token must be exactly 16 bytes")
    return token


def parse_response(data: bytes) -> tuple[bytes, int] | None:
    if len(data) != 8:
        return None
    tag, seq = struct.unpack('!4sI', data)
    if tag not in (b'RR01', b'RE01'):
        return None
    return tag, seq

File: clientRust/src/packet.rs
use std::net::IpAddr;
use std::time::{SystemTime, UNIX_EPOCH};

use hmac::{Hmac, Mac};
use sha2::Sha256;

type HmacSha256 = Hmac<Sha256>;

pub fn build_seed_packet(seq: u32, seed: &str, source_ip: IpAddr) -> [u8; 24] {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as u32;
    let time_bytes = now.to_be_bytes();

    let mut mac = HmacSha256::new_from_slice(seed.as_bytes()).expect("HMAC key error");
    mac.update(&time_bytes);
    let time_hash = mac.finalize().into_bytes();

    let ip_v6 = match source_ip {
        IpAddr::V4(v4) => v4.to_ipv6_mapped(),
        IpAddr::V6(v6) => v6,
    };

    let mut mac_ip = HmacSha256::new_from_slice(seed.as_bytes()).expect("HMAC key error");
    mac_ip.update(&time_bytes);
    mac_ip.update(&ip_v6.octets());
    let ip_hash = mac_ip.finalize().into_bytes();

    let mut packet = [0u8; 24];
    packet[0..4].copy_from_slice(b"RP01");
    packet[4..8].copy_from_slice(&seq.to_be_bytes());
    packet[8..12].copy_from_slice(&time_bytes);
    packet[12..20].copy_from_slice(&time_hash[..8]);
    packet[20..24].copy_from_slice(&ip_hash[..4]);
    packet
}

pub fn build_token_packet(seq: u32, token: &[u8; 16]) -> [u8; 24] {
    let mut packet = [0u8; 24];
    packet[0..4].copy_from_slice(b"RP01");
    packet[4..8].copy_from_slice(&seq.to_be_bytes());
    packet[8..24].copy_from_slice(token);
    packet
}

pub fn parse_response(data: &[u8]) -> Option<([u8; 4], u32)> {
    if data.len() != 8 {
        return None;
    }
    let tag: [u8; 4] = data[0..4].try_into().unwrap();
    if tag != *b"RR01" && tag != *b"RE01" {
        return None;
    }
    let seq = u32::from_be_bytes(data[4..8].try_into().unwrap());
    Some((tag, seq))
}
